manager/deallocator/deallocator.go

```diff
@@ -0,0 +1,289 @@
+package deallocator
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/docker/swarmkit/manager/state/store"
+)
+
+var (
+	resourceDeallocators = make(map[reflect.Type]resourceDeallocator)
+)
+
+func registerServiceLevelResource(resourceDeallocator resourceDeallocator) {
+	resourceDeallocators[reflect.TypeOf(resourceDeallocator.event)] = resourceDeallocator
+}
+
+// A Deallocator waits for services to fully shutdown (ie no containers left)
+// and then proceeds to deallocate service-level resources (e.g. networks),
+// and finally services themselves
+// in particular, the Deallocator should be the only place where services, or
+// service-level resources, are ever deleted!
+// it’s worth noting that this new component’s role is quite different from
+// the task reaper’s: tasks are purely internal to Swarmkit, and their status
+// is entirely managed by the system itself. In contrast, the deallocator is
+// responsible for safely deleting entities that are directly controlled by the
+// user.
+type Deallocator struct {
+	store *store.MemoryStore
+
+	// for services that are shutting down, we keep track of how many
+	// tasks still exist for them
+	services map[string]*serviceWithTaskCounts
```

Comment:
why map service IDs to both the service object AND the count? what does keeping a pointer to the service object get us, as opposed to having a `map[string]int`?

Reply:
We do need the service itself at deallocation time. Saves a lookup. I guess we could do away with that, but seems like a pretty minor choice either way?

Reply:
Sorry I didn't pay attention to this comment before, but I agree with @dperny that it is cleaner to just store service id => count. 

My suggestion: when designing data structures, its better to have as reasonably strong reason to do it a certain wait rather than it being fine either way. @wk8

Reply:
@anshulpundir : isn't saving a lookup a strong enough reason?

Reply:
One extra lookup for each service delete is not that big a overhead. But I guess you could argue either way.